resource-aggregate/cqrs/eventstore/mongodb/eventstore.go

```diff
@@ -271,234 +299,317 @@ func (s *EventStore) ensureIndex(ctx context.Context, col *mongo.Collection, ind
-		if err != nil {
-			return false, fmt.Errorf("cannot save events: %w", err)
+	aggregateID := events[0].AggregateID()
+	groupID := events[0].GroupID()
+	version := events[0].Version()
+	for idx, event := range events {
+		if event == nil {
+			return eventstore.FAIL, fmt.Errorf("invalid events[%v]('%v')", idx, event)
+		}
+		if event.AggregateID() == "" {
+			return eventstore.FAIL, fmt.Errorf("invalid events[%v].AggregateID('%v')", idx, event.AggregateID())
+		}
+		if event.GroupID() == "" {
+			return eventstore.FAIL, fmt.Errorf("invalid events[%v].GroupID('%v')", idx, event.GroupID())
+		}
+		if event.EventType() == "" {
+			return eventstore.FAIL, fmt.Errorf("invalid events[%v].EventType('%v')", idx, event.EventType())
+		}
+		if event.Version() != version+uint64(idx) {
+			return eventstore.FAIL, fmt.Errorf("invalid continues ascending events[%v].Version(%v))", idx, event.Version())
 		}
+		if idx > 0 && event.IsSnapshot() {
+			return eventstore.FAIL, fmt.Errorf("inner snapshot for events[%v] is not supported", idx)
+		}
+		if event.AggregateID() != aggregateID {
+			return eventstore.FAIL, fmt.Errorf("invalid events[%v].AggregateID('%v') != events[0].AggregateID('%v')", idx, event.AggregateID(), aggregateID)
+		}
+		if event.GroupID() != groupID {
+			return eventstore.FAIL, fmt.Errorf("invalid events[%v].GroupID('%v') != events[0].GroupID('%v')", idx, event.GroupID(), groupID)
+		}
+
 	}
 
+	col := s.client.Database(s.DBName()).Collection(getEventCollectionName())
 	if events[0].Version() == 0 {
-		concurrencyException, err = s.SaveSnapshotQuery(ctx, collectionID, aggregateID, 0)
+		doc, err := makeDBDoc(events, s.dataMarshaler)
 		if err != nil {
-			return false, fmt.Errorf("cannot save events without snapshot query for version 0: %w", err)
+			return eventstore.FAIL, fmt.Errorf("cannot insert first events('%v'): %w", events, err)
 		}
-		if concurrencyException {
-			return concurrencyException, nil
+		_, err = col.InsertOne(ctx, doc)
+		if err != nil {
+			if IsDup(err) {
+				return eventstore.ConcurrencyException, nil
+			}
+			return eventstore.FAIL, fmt.Errorf("cannot insert first events('%v'): %w", events, err)
 		}
+		return eventstore.OK, nil
 	}
-
-	if len(events) > 1 {
-		return s.saveEvents(ctx, col, collectionID, aggregateID, events)
+	if events[0].IsSnapshot() {
+		return s.saveSnapshot(ctx, events)
 	}
-	return s.saveEvent(ctx, col, collectionID, aggregateID, events[0])
+
+	return s.saveEvent(ctx, col, events)
 }
 
-func (s *EventStore) SaveSnapshot(ctx context.Context, collectionID string, aggregateID string, ev eventstore.Event) (concurrencyException bool, err error) {
-	concurrencyException, err = s.Save(ctx, collectionID, aggregateID, []eventstore.Event{ev})
-	if !concurrencyException && err == nil {
-		return s.SaveSnapshotQuery(ctx, collectionID, aggregateID, ev.Version())
+func (s *EventStore) saveSnapshot(ctx context.Context, events []eventstore.Event) (status eventstore.SaveStatus, err error) {
```

Comment:
Verify if you can load just last x events until you reach latest snapshot. If yes, don't create new document everytime when snapshot is created.